### ISSUE TITLE
support for dfpwm audio format

### ISSUE THREAD

**Issue by charlie-sans** [Yellow-Dog-Man/Resonite-Issues#2831]
### Is your feature request related to a problem? Please describe.

i really want to be able to ditch wav / ogg for dfpwm but i can't import them into the game without just importing the raw files without being in dfpwm

### Describe the solution you'd like

being able to drag and drop dfwpm files into resonite and they show up as a audio file like wav, flacc and ogg.

### Describe alternatives you've considered

i was thinking of making a mod that does this but i have no modding experience.

### Additional Context

found something that might help with research https://github.com/BlueAmulet/LionRay

### Requesters

charlie_san on discord and resonite

**Comment by Frooxius** [comment]
Why do you want to use this format over other supported formats that we already support? What benefits they provide?

How commonly is this format used?

Supporting a new format can be a pretty significant long term commitment on our end, so we'd need some more rationale for supporting this.

**Comment by charlie-sans** [comment]
dfpwm is mostly used with the computercraft mod for minecraft and some other projects but it's pretty common for people to host dfpwm converters or make ones that go on github for people to use.

with the size of dfpwm files from what i have seen atleast. when bringing in the size of the audio. dfpwm does win in some parts but i would recommend looking at https://wiki.vexatos.com/dfpwm for more infomation regarding dfpwm.

**Comment by Frooxius** [comment]
Hmm... This sounds like it's very niche use-case. We'd probably not commit towards supporting this format based on that, unless there's significant interest or benefits.

**Comment by iamgreaser** [comment]
I am the original author of this codec.

Something to note is that the notes on that page are out of date by about a year. The usual version *actually* used by people is DFPWM1a (this change happened around the time of one of the BTM conventions). It typically runs at 48000 Hz if I recall correctly, and sounds notably less "mushy" than the original DFPWM1.

The argument in favour of supporting this codec is that it is very simple and quick to decode. I believe asie made a player that could play in realtime on a(n emulated) 12 MHz 286 though I don't recall the exact number in this case.

But I would be more in favour of supporting the import of Opus audio as one of the audio formats, similar to how Ogg Vorbis is supported (typically Opus is itself encapsulated in an Ogg container). I did some casual testing back in the day and it definitely outperformed DFPWM1a in terms of being able to compress to a low bitrate and sound good.

In the meantime, there is a small chance that Resonite may *accidentally* end up with support for it once #1476 happens. It's actually supported by ffmpeg now. No, I wasn't responsible for that port.

**Comment by charlie-sans** [comment]
hey greaser! 

makes sense that the page is out of date, though it was the only good one i could find.
if libvlc does get support and updated then it might be added which is nice.

i manly wanted it so that people could import short sound files without it being 4 something gigs per 1 seccond (just as a joke though).

**Comment by Frooxius** [comment]
Yeah, Opus is definitely something we want to support for direct audio playback and we're a lot more likely invest time into that one.

If there's a support for this codec in FFMpeg, that makes it a bit easier, because it offloads the support, but we'd still need to put substantial effort to use FFMpeg as audio decoder. 

@charlie-sans You can import short sound files as OGG Vorbis, which will be pretty small. I don't really see this as better alternative at this point, over something a lot more widespread.

**Comment by charlie-sans** [comment]
> But I would be more in favour of supporting the import of Opus audio as one of the audio formats, similar to how Ogg Vorbis is supported (typically Opus is itself encapsulated in an Ogg container). I did some casual testing back in the day and it definitely outperformed DFPWM1a in terms of being able to compress to a low bitrate and sound good.

that makes sense, having support for Opus audio would be better for people. i think having Opus would be better then dfpwm in terms of support because if your correct with libvlc then we could get most or all of the good ones. 

though it would be nice if mp3 was a audio file instead of a video player so we could use that aswell for music instead of wav or flacc.
